Flywheel

crn2267 opened the pull request:
Basic functionality of a Flywheel class. Takes a motor group, robot config, and PID in the constructor; has async RPM setting, sync spin / stop commands.

crn2267 committed:
Initial commit.

crn2267 committed:
Added async functionality to Flywheel.

superrm11 commented on core/include/subsystems/flywheel.h:
```diff
@@ -0,0 +1,29 @@
+  PID pid;
+  bool inRPMLoop = false;
+  double RPM = -1.0;
+  thread rpmThread;
```
consider changing the thread to a task. Since PID requires a delay to operate, it'll be less wasted CPU time during that delay since the task scheduler takes care of it

superrm11 commented on core/include/subsystems/flywheel.h:
```diff
@@ -0,0 +1,29 @@
+class Flywheel{
+  public:
+
+  Flywheel(motor_group &motors, robot_specs_t &config, PID &pid);
```
PID object can be created and stored in the class, you can pass in a PID::pid_config_t reference and create it in the constructor

Also, it doesn't look like the robot_config_t parameter is used in the class

superrm11 commented on core/src/subsystems/flywheel.cpp:
```diff
@@ -0,0 +1,71 @@
+
+// Spin motors using voltage; defaults forward at 12 volts
+// Speed is between -1 and 1.
+void Flywheel::spin(double speed, directionType dir){
```
rename to spin_raw or spin_pct? something to communicate it's not set by a control loop

superrm11 commented on core/src/subsystems/flywheel.cpp:
```diff
@@ -0,0 +1,71 @@
+}
+
+// starts / restarts RPM thread at new value
+void Flywheel::spinRPM(int inputRPM) {
```
If someone puts this in the main loop, it'll be starting/stopping the thread over and over again, which could cause issues. Consider following something like [the task management in Lift.h](https://github.com/RIT-VEX-U/Core/blob/9a9c7c9806ded60ed169a6e8e3d4fc26bc2eeae8/include/subsystems/lift.h#L63). It starts the task in the constructor and only changes the target in set() methods.

You can also replace the getPID and getMotors with a single get_target method in this case

superrm11 commented on core/src/subsystems/flywheel.cpp:
```diff
@@ -0,0 +1,71 @@
+#include "../core/include/subsystems/flywheel.h"
+#include "../core/include/utils/pid.h"
+#include "vex.h"
+#include <thread>
```
Vex doesn't  support std::thread, only vex::thread and vex::task

superrm11 reviewed (changes requested):
Remember to document the code with block comments so we can generate documentation later on. [Here's the style guide for doing that](https://docs.google.com/document/d/1FIJRD8D6g59M6G-C9oESVdkim6jV2vFhITjwuyI32jg/edit?usp=sharing).

Nice progress

superrm11 commented:
Also, I'll go ahead and push the FeedForward class to this repo; I realize now I recommended PID+Feedforward for the control loop but didn't actually supply the class...

crn2267 committed:
Handling McGee's requests, pt. 1.

crn2267 committed:
Merge

crn2267 committed:
Merge pt. ???

crn2267 committed:
Fixed merge conflicts in project files (.v5code, .json)

crn2267 committed:
v5code keeps breaking AAAAAAAAAAAAAA

crn2267 committed:
.json files make me cringe.

crn2267 committed:
Added documentation, fixed async stuff.

crn2267 committed:
Added 'bang bang'.

crn2267 committed:
Changed the Flywheel constructor to take a pid *config* pointer instead of a PID pointer.

crn2267 committed:
Documentation and organization improvements, added stubs for RPM control test functions.

superrm11 commented on core/src/subsystems/flywheel.cpp:
```diff
@@ -17,56 +19,129 @@
+*/
+void Flywheel::spinRPM(int inputRPM) {
+  // only run if the RPM is different or it isn't already running
+  if(inputRPM != RPM || !taskRunning) {
```
While this would probably work for instant changes in RPM setpoint, if someone wanted to ramp to a value the task scheduler would probably have a stroke creating and destroying the task every  time it loops. See above for my recommendation about creating the task once in the constructor

superrm11 replied:
One thing I like to do is design it as if it wasn't an asynchronous class, and you had some sort of `update()` method that did the calculations and set the motors that'd the programmer would call in the main loop. Then for the task, all you'd do is have that `update()` function with a delay

crn2267 replied:
Fixed (probably); Dana says it _looks_ like it will work.

superrm11 reviewed (commented):
Love the changelog at the top btw!

crn2267 committed:
Updated async to not 'have a stroke'.

superrm11 commented:
Looks good to me. Let me know how it tests and I'll merge it in

crn2267 committed:
Changes made during testing, futile attempts at debugging.

crn2267 committed:
Changes made during testing, futile attempts at debugging.

cowsed committed:
Commented out old printf s
Terminal is cleaner now

cowsed committed:
Flywheel PID + Feedforward, Feedforward.
Functions and configs

cowsed committed:
Set up reverses for drive wheels

cowsed committed:
Take Back Half control and tuning.
Right now, the tbh gain parameter is hardcoded into the flywheel definition, this should be extracted so we can use the same code with different configs for different flywheels.

cowsed committed:
Created multiple ways to construct a Flywheel each corresponding to a different control method
Flywheel now chooses which spinRPMtask to use based on this

cowsed committed:
Formalized TBH Control in robot config
Cleaned up opcontrol to remove all testing stuff

cowsed committed:
Revert "Formalized TBH Control in robot config"

This reverts commit 032ecfddbaaa20af02931ae0630a58e037e32a70.

superrm11 commented:
Some final changes I'll suggest:
- changing RPM variable from a `double` to a `std::atomic<double>` type to avoid a possible race condition
- Resetting files not related to the Flywheel to how they were before branching off (to prevent merge conflicts)
  - Ideally the only changes (I think) should be to flywheel.h, flywheel.cpp, the .v5code/compile commands, and core.h
- Adding an extra asterisk to the block comments, so our auto-gen documentation tool can recognize it. Also, copying over the documentation to the header file

Example:
```C++
/*
 *
 */

// To

/**
  *
  */
```

cowsed commented:
In terms of resetting files, do you want me to get rid of the flywheel tuning parameters from robot_config and store those somewhere else or should I leave those in place. (I'll get rid of all the other testing nonsense in robot_config though)

superrm11 commented:
robot-config can stay actually, since it has tunings

cowsed committed:
Reset non-flywheel related files to their initial state

cowsed committed:
removed spooky flywheel adjacent but not related files that will come back to haunt us later

cowsed committed:
Updated documentation and included it in the header file

superrm11 committed:
Merge branch 'main' into Flywheel
